include/verilated_timing.h

```diff
@@ -0,0 +1,367 @@
+
+    // METHODS
+    // Move coroutines waiting on the specified events to the array
+    void move(const VerilatedEventSet& events, VerilatedCoroutineArray& coros) {
+        auto range = m_eventSetsToCoros.equal_range(events);
+        for (auto it = range.first; it != range.second; ++it) coros.push_back(it->second);
+        m_eventSetsToCoros.erase(range.first, range.second);
+    }
+
+public:
+    ~VerilatedEventToCoroMap() {
+        for (auto& p : m_eventSetsToCoros) { p.second.destroy(); }
+    }
+
+    // Move coroutines waiting on the specified event to the array
+    void move(VerilatedEvent* event, VerilatedCoroutineArray& coros) {
+        auto range = m_eventsToEventSets.equal_range(event);
+        for (auto it = range.first; it != range.second; ++it) move(it->second, coros);
+    }
+
+    // Are there coroutines waiting on this set of events?
+    bool contains(const VerilatedEventSet& events) {
+        auto range = m_eventSetsToCoros.equal_range(events);
+        return range.first != range.second;
+    }
+
+    // Insert a coroutine waiting on the specified events
+    void insert(const VerilatedEventSet& events, std::coroutine_handle<> coro) {
+        for (auto event : events) m_eventsToEventSets.insert(std::make_pair(event, events));
+        m_eventSetsToCoros.insert(std::make_pair(events, coro));
+    }
+
+    // Erase all coroutines waiting on the given event
+    void erase(VerilatedEvent* event) {
+        auto range = m_eventsToEventSets.equal_range(event);
+        for (auto it = range.first; it != range.second; ++it) m_eventSetsToCoros.erase(it->second);
+    }
+};
+
+//=============================================================================
+// VerilatedEventDispatcher
+// Object that manages event variables and suspended coroutines waiting on those event vars.
+// When an event gets triggered, the corresponding coroutine isn't resumed immediately. This is to
+// allow for all processes to get to their timing controls, so that order of execution of processes
+// doesn't matter as much. Coroutines are resumed later by the 'resumeTriggered` function. The only
+// exception to this is when an event gets triggered twice – then all coroutines up to the first
+// trigger get resumed. This is neccessary so we don't lose track of events getting triggered and
+// the order they're triggered in.
+class VerilatedEventDispatcher final {
+private:
+    // MEMBERS
+    VerilatedEventToCoroMap m_eventsToCoros;  // Mapping of events to coroutines
+    std::vector<VerilatedEvent*> m_triggeredEvents;  // Events triggered in the current time slot
+    VerilatedEventSet m_eventsToReset;  // Events to be reset at the start of a time slot
+    VerilatedCoroutineArray
+        m_readyCoros;  // Coroutines ready for resumption in the current time slot
+    VerilatedMutex m_mutex;  // Protects m_eventSetsToCoros, m_triggeredEvents, m_eventsToCoros,
+                             // and m_readyCoros
```

Comment:
```suggestion
    mutable VerilatedMutex m_mutex;  // Protects m_eventSetsToCoros, m_triggeredEvents, 
                             // m_eventsToCoros, and m_readyCoros
```

Mutable so can put const on appropriate methods.